### ISSUE TITLE
no password or OTP prompt on Openwrt compile

### ISSUE THREAD

**Issue by maurerr** [adrienverge/openfortivpn#360]
Hello,

I've compiled a openfrotivpn for openwrt-x86 using a makefile from https://github.com/adrienverge/openfortivpn/issues/80
when i try to connect to fortigate vpn server i get no password or OTP prompt:

root@OpenWrt:~# openfortivpn -vvv test-fgt.domain.com:443 -u test.user
WARN:   Bad port in config file: "0".
DEBUG:  Loaded config file "/etc/openfortivpn/config".

If I enter the password and later the OTP the connection is established so only a minor bug  - no critical one.
I just want to know if it's openwrt or openfortivpn related.
You can use this OVA in virtualbox (recommended) or vmware (untested) to reproduce the issue:
https://gofile.io/?c=8JgobS

**Comment by DimitriPapadopoulos** [comment]
The config file is incorrect as it hasn't been changed from the template, hence the warning messages. That shouldn't be an issue though.

The part of the code handling user input is here:
https://github.com/adrienverge/openfortivpn/blob/a5ab08a/src/userinput.c#L26-L51
As you can see the prompt is printed by a mere _printf_.

I don't know if it's an OpenWrt issue, but I know this code does work on Linux, FreeBSD or macOS.

Perhaps you need to flush the buffer after the _printf_ call:
```fflush(stdout);```

**Comment by maurerr** [comment]
i recompiled it adding is:
`...
        printf("\n");
        fflush(stdout);
}
`
 but no change 

```
root@OpenWrt:~# openfortivpn -vvv 10.1.1.1:443 -u test.user
DEBUG:  Loaded config file "/etc/openfortivpn/config".
```

**Comment by DimitriPapadopoulos** [comment]
Strange, I have no clue at this point. Other parts of the code such as logging use _printf_ and _fflush_ and logging seems to be working for you.
https://github.com/adrienverge/openfortivpn/blob/a5ab08a/src/log.c#L90-L110

**Comment by DimitriPapadopoulos** [comment]
If you add:
````printf("FOO\n")````
before 
```printf("%s", prompt);```
do you see `FOO` printed?

If not, I suggest you move ``printf("FOO\n")`` up into the functions that call _read_password_. It should start printing before you reach _main_. If not I really don't know and it's probably an OpenWrt subtlety!

**Comment by maurerr** [comment]
we're moving forward :)
FOO is printed:
root@OpenWrt:~# openfortivpn test-fgt.my.to:443 -u test-user
FOO

but still no password prompt

**Comment by DimitriPapadopoulos** [comment]
Try changing:
```printf("%s", prompt);```
to:
```printf("FOO\n%s", prompt);```
What do you see?

Then where exactly had you add the following?
```... printf("\n"); fflush(stdout); }```

It should  have printed something if just after:
```printf("%s", prompt);```

**Comment by DimitriPapadopoulos** [comment]
The other possibility is that the `prompt` is an empty string - but that would be weird!

**Comment by maurerr** [comment]
so i've added printf("FOO\n%s", prompt); but no FOO this time
`root@OpenWrt:~# openfortivpn test-fgt.my.to:443 -u test-user
WARN:   Bad port in config file: "0".`

earlier I've added `fflush(stdout);` right before the last line of userinput.c

**Comment by DimitriPapadopoulos** [comment]
It does make sense to `fflush(stdout)` after `printf()` because we need immediate output before starting to read the password or OTP. This is specially true if the printed string lacks a line break `\n` - line breaks usually implicitly flush the `stdout` buffer.

This is clearly documented in the [POSIX documentation](http://pubs.opengroup.org/onlinepubs/009695399/functions/fflush.html):
> ### EXAMPLES
> 
> #### Sending Prompts to Standard Output
> 
> The following example uses [_printf_()](http://pubs.opengroup.org/onlinepubs/9699919799/functions/printf.html) calls to print a series of prompts for information the user must enter from standard input. The _fflush_() calls force the output to standard output. The _fflush_() function is used because standard output is usually buffered and the prompt may not immediately be printed on the output or terminal. The [_getline_()](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getline.html) function calls read strings from standard input and place the results in variables, for use later in the program.
> ```
> char *user;
> char *oldpasswd;
> char *newpasswd;
> ssize_t llen;
> size_t blen;
> struct termios term;
> tcflag_t saveflag;
> 
> printf("User name: ");
> fflush(stdout);
> blen = 0;
> llen = getline(&user, &blen, stdin);
> user[llen-1] = 0;
> tcgetattr(fileno(stdin), &term);
> saveflag = term.c_lflag;
> term.c_lflag &= ˜ECHO;
> tcsetattr(fileno(stdin), TCSANOW, &term);
> printf("Old password: ");
> fflush(stdout);
> blen = 0;
> llen = getline(&oldpasswd, &blen, stdin);
> oldpasswd[llen-1] = 0;
> 
> printf("\nNew password: ");
> fflush(stdout);
> blen = 0;
> llen = getline(&newpasswd, &blen, stdin);
> newpasswd[llen-1] = 0;
> term.c_lflag = saveflag;
> tcsetattr(fileno(stdin), TCSANOW, &term);
> free(user);
> free(oldpasswd);
> free(newpasswd);
> ```

What's puzzling me is that the `stdout` buffer should neverthless be flushed later on, as soon as a string containing a line break `\n` is printed. So the prompt should eventually be printed, even asynchronously and after you enter the password. Don't you see that?

**Comment by maurerr** [comment]
you're right.
i do see it later on if I enter the password or just press "Enter"

**Comment by mrbaseman** [comment]
the fix for this went into 1.8.0 which has been released recently
